tests/fixtures/test_auth.py

```diff
@@ -93,3 +93,181 @@ def test_login(client, users):
+        'emailAddress': 'test@digital.gov.au', 'password': 'testpassword'
+    }), content_type='application/json')
+    assert res.status_code == 200
+
+    res = client.post('/2/generate-api-key')
+    assert res.status_code == 200
+
+    data = json.loads(res.get_data())
+    key = data['key']
+
+    res = client.get('/2/logout')
+    assert res.status_code == 200
+
+    res = client.get('/2/ping')
+    data = json.loads(res.get_data(as_text=True))
+    assert not data['isAuthenticated']
+
+    res = client.get('/2/ping', headers={'X-Api-Key': key})
+    data = json.loads(res.get_data(as_text=True))
+    assert data['isAuthenticated']
+
+
+def test_api_key_authentication_fails_on_non_api_key_resource(client, users):
+    res = client.post('/2/login', data=json.dumps({
+        'emailAddress': 'test@digital.gov.au', 'password': 'testpassword'
+    }), content_type='application/json')
+    assert res.status_code == 200
+
+    res = client.post('/2/generate-api-key')
+    assert res.status_code == 200
+
+    data = json.loads(res.get_data())
+    key = data['key']
+
+    res = client.get('/2/logout')
+    assert res.status_code == 200
+
+    res = client.get('/2/_protected', headers={'X-Api-Key': key})
+    assert res.status_code == 401
+
+
+def test_api_key_authentication_fails_supplier_user(client, supplier_user):
+    res = client.post('/2/login', data=json.dumps({
+        'emailAddress': 'j@examplecompany.biz', 'password': 'testpassword'
+    }), content_type='application/json')
+    assert res.status_code == 200
+
+    res = client.get('/2/ping')
+    data = json.loads(res.get_data(as_text=True))
+    assert data['isAuthenticated']
+
+    res = client.post('/2/generate-api-key')
+    assert res.status_code == 403
+
+
+def test_api_key_authentication_fails_bad_header(client, users):
+    res = client.post('/2/login', data=json.dumps({
+        'emailAddress': 'test@digital.gov.au', 'password': 'testpassword'
+    }), content_type='application/json')
+    assert res.status_code == 200
+
+    res = client.post('/2/generate-api-key')
+    assert res.status_code == 200
+
+    data = json.loads(res.get_data())
+    key = data['key']
+
+    res = client.get('/2/logout')
+    assert res.status_code == 200
+
+    res = client.get('/2/ping')
+    data = json.loads(res.get_data(as_text=True))
+    assert not data['isAuthenticated']
+
+    res = client.get('/2/ping', headers={'X-Apikey': key})
+    data = json.loads(res.get_data(as_text=True))
+    assert not data['isAuthenticated']
+
+    res = client.get('/2/ping', headers={'X-Api-key': 'badkey'})
+    assert res.status_code == 403
+
+
+def test_api_key_revocation(client, users):
+    res = client.post('/2/login', data=json.dumps({
+        'emailAddress': 'test@digital.gov.au', 'password': 'testpassword'
+    }), content_type='application/json')
+    assert res.status_code == 200
+
+    res = client.post('/2/generate-api-key')
+    assert res.status_code == 200
```

Comment:
Tests are good.  Just wondering if we need to repeat some of these asserts if we're already checking them elsewhere?

Reply:
I think it helps to quickly pin point where a failing test is actually failing. It can also avoid cases where a failure in setup can lead to false positives (although I don't think this would describe any API key test).